core/src/selection/selectionQuery.cpp

```diff
@@ -23,17 +23,33 @@ QueryType SelectionQuery::type() const {
 void SelectionQuery::process(const View& _view, const FrameBuffer& _framebuffer, const MarkerManager& _markerManager,
                              const TileManager& _tileManager, const Labels& _labels, std::vector<SelectionColorRead>& _colorCache) const {
 
-    glm::vec2 windowCoordinates = _view.normalizedWindowCoordinates(m_position.x, m_position.y);
+    float radius = m_radius * _view.pixelScale();
+    glm::vec2 windowCoordinates = _view.normalizedWindowCoordinates(m_position.x - radius, m_position.y + radius);
+    glm::vec2 windowSize = _view.normalizedWindowCoordinates(m_position.x + radius, m_position.y - radius) - windowCoordinates;
 
     GLuint color = 0;
 
     auto it = std::find_if(_colorCache.begin(), _colorCache.end(), [=](const auto& _colorRead) {
-        return m_position == _colorRead.position;
+        return (m_position == _colorRead.position) && (m_radius == _colorRead.radius);
     });
 
     if (it == _colorCache.end()) {
-        color = _framebuffer.readAt(windowCoordinates.x, windowCoordinates.y);
-        _colorCache.push_back({color, m_position});
+        // Find the first non-zero color nearest to the position and within the selection radius.
+        auto rect = _framebuffer.readRect(windowCoordinates.x, windowCoordinates.y, windowSize.x, windowSize.y);
+        float minDistance = std::fmin(rect.width, rect.height);
+        float hw = static_cast<float>(rect.width) / 2.f, hh = static_cast<float>(rect.height) / 2.f;
+        for (int32_t row = 0; row < rect.height; row++) {
+            for (int32_t col = 0; col < rect.width; col++) {
+                uint32_t sample = rect.pixels[row * rect.width + col];
```

Comment:
also we should break the loop if `dist = 1`

Also searching from the center might be more efficient more larger rect ?
ie with such a pattern (where number correspond to loop iteration):

```
3333333
3222223
3211123
3210123
3211123
3222223
3333333
```

The algo would need to skip location outside the rect,
The loop can stop when the min distance on the next iteration (=`n+1` above) is more than the hit min distance.